### DISCUSSION TITLE
Configuration file option `exclude_paths` doesn't seem to have any effect

### DISCUSSION THREAD

**Discussion by Rylon** [ansible/ansible-lint#1561] [Q&A] [answered]
##### Summary

I noticed that a path I added to `exclude_paths` in the `ansible-lint.yml` configuration file wasn't taking effect.

##### Issue Type

- Bug Report

##### Ansible and Ansible Lint details

```console
$ ansible --version
ansible 2.10.8

$ ansible-lint --version
ansible-lint 5.0.3 using ansible 2.10.8
```

- ansible installation method: pip
- ansible-lint installation method: pip

##### OS / ENVIRONMENT

Docker running Debian 10 image, Python 3.8.5

##### STEPS TO REPRODUCE

I have the following in my `exclude_paths` which excludes third-party roles (which we vendor inside our own repository) from being linted.

```console
exclude_paths:
  - roles-vendored/
```

If I lint a playbook which includes one of these vendored roles, the linter is still trying to lint those files from the vendored role, despite it being in the exclude path.

If I pass this option via the command line flag, the behaviour works as expected:

```console
ansible-lint --exclude roles-vendored/
```

##### Desired Behaviour

No files from the `exclude_paths` of the configuration file should be considered for linting.

##### Actual Behaviour

Files inside the excluded path are linted, unless the `--exclude` command line flag is used.

**Comment by devjatkin** [comment, marked as answer]
Hello @Rylon,

Do you still experience that issue? If yes, could you please provide a bit more details by passing "-vvv" argument in command line, like that:
`ansible-lint -vvv `

Thank you in advance

**Reply by Rylon** [reply]
Hi, here's the output with the exclude added to the command-line, which works correctly:

```
$ ansible-lint --exclude roles-vendored/ --offline -c .ansible-lint/ansible-lint.yml roles-vendored/geerlingguy.logstash/tasks/config.yml
.ansible-lint/ansible-lint.yml roles-vendored/geerlingguy.logstash/tasks/config.yml 
Added ANSIBLE_ROLES_PATH=~/.ansible/roles:/usr/share/ansible/roles:/etc/ansible/roles:roles
Loading custom .yamllint config file, this extends our internal yamllint config.
INFO     Discovering files to lint: git ls-files -z *.yaml *.yml
```

If I remove the `--exclude` option, you can see the fake errors I added to the file causing the linting to fail:

```
ansible-lint --offline -c .ansible-lint/ansible-lint.yml roles-vendored/geerlingguy.logstash/tasks/config.yml 
Added ANSIBLE_ROLES_PATH=~/.ansible/roles:/usr/share/ansible/roles:/etc/ansible/roles:roles
Loading custom .yamllint config file, this extends our internal yamllint config.
INFO     Discovering files to lint: git ls-files -z *.yaml *.yml
WARNING  Listing 2 violation(s) that are fatal
DE001: Use pure YAML style (dictionaries), not the older style key=value
roles-vendored/geerlingguy.logstash/tasks/config.yml:14 - name: test=bob

parser-error: no module/action detected in task.
roles-vendored/geerlingguy.logstash/tasks/config.yml:14

You can skip specific rules or tags by adding them to your configuration file:
# .ansible-lint
warn_list:  # or 'skip_list' to silence them completely
  - DE001  # Use pure YAML style (dictionaries), not the older style key=value
  - parser-error  # AnsibleParserError

Finished with 2 failure(s), 0 warning(s) on 1 files.
```

Content of `.ansible-lint/ansible-lint.yml`:

```
---
exclude_paths:
  - roles-vendored/
rulesdir:
  - ./rules/
use_default_rules: true
verbosity: 1
```

**Reply by devjatkin** [reply]
your configuration file will exclude path .ansible-lint/roles-vendored/ what you can double check with higher level of verbosity
try to change config in way: - ../roles-vendored/

**Reply by Rylon** [reply]
Ah interesting, yeah it started working when I changed the config file to `- ../roles-vendored`! The previous behaviour this was relative to the working directory, so I guess now it is relative to the location of the configuration file?

**Comment by ssbarnea** [comment]
I am almost sure that the config file is not loaded. It works and we have tests relying on it to work.

**Reply by Rylon** [reply]
Hi,

The config file seems to be working for the other settings we change, just not for this option. I have to set it on the command line for it to work.

**Comment by greg-hellings** [comment]
So I'm now encountering this manifestation of this bug:

* Passing `--exclude roles/system_repositories/molecule` on the command line properly excludes the folder
* Setting `exclude_paths: ['roles/system_repositories/molecule']` in the file does not produce an effect.

Setting double verbosity seems to expose a possible reason why. The values from the two sources are not being parsed the same.

When passing with the config file only, I get this result:

```
DEBUG    Options: Namespace(cache_dir='/var/home/ghelling/.cache/ansible-lint/97e51a', colored=True, configured=True, cwd=PosixPath('/var/home/ghelling/src/ansible_collections/oasis_roles/system'), display_relative_path=True, exclude_paths=['/var/home/ghelling/src/ansible_collections/oasis_roles/system/tests/roles/system_repositories/molecule'], lintables=[], listrules=False, listtags=False, parseable=False, parseable_severity=False, quiet=False, rulesdirs=['/var/home/ghelling/venv/tools3/lib64/python3.9/site-packages/ansiblelint/rules'], skip_list=['204', 'yaml'], tags=[], verbosity=2, warn_list=['experimental', 'role-name'], kinds=[{'jinja2': '**/*.j2'}, {'jinja2': '**/*.j2.*'}, {'inventory': '**/inventory/**.yml'}, {'requirements': '**/meta/requirements.yml'}, {'galaxy': '**/galaxy.yml'}, {'reno': '**/releasenotes/*/*.{yaml,yml}'}, {'playbook': '**/playbooks/*.{yml,yaml}'}, {'playbook': '**/*playbook*.{yml,yaml}'}, {'role': '**/roles/*/'}, {'tasks': '**/tasks/**/*.{yaml,yml}'}, {'handlers': '**/handlers/*.{yaml,yml}'}, {'vars': '**/{host_vars,group_vars,vars,defaults}/**/*.{yaml,yml}'}, {'meta': '**/meta/main.{yaml,yml}'}, {'yaml': '.config/molecule/config.{yaml,yml}'}, {'requirements': '**/molecule/*/{collections,requirements}.{yaml,yml}'}, {'yaml': '**/molecule/*/{base,molecule}.{yaml,yml}'}, {'requirements': '**/requirements.yml'}, {'playbook': '**/molecule/*/*.{yaml,yml}'}, {'yaml': '**/{.ansible-lint,.yamllint}'}, {'yaml': '**/*.{yaml,yml}'}, {'yaml': '**/.*.{yaml,yml}'}], mock_modules=[], mock_roles=[], loop_var_prefix=None, var_naming_pattern=None, offline=False, project_dir='tests', extra_vars=None, enable_list=[], skip_action_validation=True, rules={}, format='rich', progressive=False, rulesdir=[], use_default_rules=True, config_file='/var/home/ghelling/src/ansible_collections/oasis_roles/system/tests/ansible-lint.yml', version=False)
```

When passing from the command line, instead, I get this

```
DEBUG    Options: Namespace(cache_dir='/var/home/ghelling/.cache/ansible-lint/97e51a', colored=True, configured=True, cwd=PosixPath('/var/home/ghelling/src/ansible_collections/oasis_roles/system'), display_relative_path=True, exclude_paths=[PosixPath('/var/home/ghelling/src/ansible_collections/oasis_roles/system/roles/system_repositories/molecule')], lintables=[], listrules=False, listtags=False, parseable=False, parseable_severity=False, quiet=False, rulesdirs=['/var/home/ghelling/venv/tools3/lib64/python3.9/site-packages/ansiblelint/rules'], skip_list=['204', 'yaml'], tags=[], verbosity=2, warn_list=['experimental', 'role-name'], kinds=[{'jinja2': '**/*.j2'}, {'jinja2': '**/*.j2.*'}, {'inventory': '**/inventory/**.yml'}, {'requirements': '**/meta/requirements.yml'}, {'galaxy': '**/galaxy.yml'}, {'reno': '**/releasenotes/*/*.{yaml,yml}'}, {'playbook': '**/playbooks/*.{yml,yaml}'}, {'playbook': '**/*playbook*.{yml,yaml}'}, {'role': '**/roles/*/'}, {'tasks': '**/tasks/**/*.{yaml,yml}'}, {'handlers': '**/handlers/*.{yaml,yml}'}, {'vars': '**/{host_vars,group_vars,vars,defaults}/**/*.{yaml,yml}'}, {'meta': '**/meta/main.{yaml,yml}'}, {'yaml': '.config/molecule/config.{yaml,yml}'}, {'requirements': '**/molecule/*/{collections,requirements}.{yaml,yml}'}, {'yaml': '**/molecule/*/{base,molecule}.{yaml,yml}'}, {'requirements': '**/requirements.yml'}, {'playbook': '**/molecule/*/*.{yaml,yml}'}, {'yaml': '**/{.ansible-lint,.yamllint}'}, {'yaml': '**/*.{yaml,yml}'}, {'yaml': '**/.*.{yaml,yml}'}], mock_modules=[], mock_roles=[], loop_var_prefix=None, var_naming_pattern=None, offline=False, project_dir='tests', extra_vars=None, enable_list=[], skip_action_validation=True, rules={}, format='rich', progressive=False, rulesdir=[], use_default_rules=True, config_file='/var/home/ghelling/src/ansible_collections/oasis_roles/system/tests/ansible-lint.yml', version=False)
```

The values coming from the config file are read as relative to the config file and not relative to the CWD. Personally I would say that this is a bug. Or, at the least, it should be documented that the paths are relative to the config file if they're in the file or to CWD if they're found on the command line.

As an additional interesting note, the values read from the file are handled as strings whereas the values pulled from the command line are instances of the class PosixPath. I'm not sure that this means much, but it is worth noting that the file and CLI are not being handled the same.

Currently, also, we do prefix string matching. It might be worth considering use of the [fnmatch](https://docs.python.org/3/library/fnmatch.html) module, instead, so that we can support things like glob paths. Currently, exclude_paths does not have very good support for glob expressions.

I can probably work up a PR to at least provide the glob abilities and either fix or document the differences between the CLI and the config_file values for this argument.
